lib/heuristics.ts

```diff
@@ -103,3 +18,208 @@ export function checkHeuristicPatterns() {
     }
     return { patterns, snippets };
 }
+
+export function getActionablePopups(): PopupData[] {
+    const popups = getPotentialPopups();
+    const result = popups.reduce((acc, popup) => {
+        const popupText = popup.text?.trim();
+        if (popupText) {
+            const { patterns } = checkHeuristicPatterns(popupText);
+            if (patterns.length > 0) {
+                const { rejectButtons, otherButtons } = classifyButtons(popup.buttons);
+                if (rejectButtons.length > 0) {
+                    acc.push({
+                        ...popup,
+                        rejectButtons,
+                        otherButtons,
+                    });
+                }
+            }
+        }
+        return acc;
+    }, [] as PopupData[]);
+    return result;
+}
+
+function classifyButtons(buttons: ButtonData[]): { rejectButtons: ButtonData[]; otherButtons: ButtonData[] } {
+    const rejectButtons = [];
+    const otherButtons = [];
+    for (const button of buttons) {
+        if (isRejectButton(button.text)) {
+            rejectButtons.push(button);
+        } else {
+            otherButtons.push(button);
+        }
+    }
+    return {
+        rejectButtons,
+        otherButtons,
+    };
+}
+
+function isRejectButton(buttonText: string): boolean {
+    if (!buttonText) {
+        return false;
+    }
+    const cleanedButtonText = cleanButtonText(buttonText);
+    return (
+        !NEVER_MATCH_PATTERNS.some((p) => p.test(cleanedButtonText)) &&
+        REJECT_PATTERNS.some((p) => (p instanceof RegExp && p.test(cleanedButtonText)) || p === cleanedButtonText)
+    );
+}
+
+function cleanButtonText(buttonText: string): string {
```

Comment:
I think I'd like to see tests for all these utility functions, otherwise I think this code will be very difficult to change without regressions.